packages/nep141-erc20/src/utils/index.js

```diff
@@ -0,0 +1,84 @@
+import Web3 from 'web3'
+import { getEthProvider } from '@near-eth/client/dist/utils'
+
+/**
+ * Search for a dropped and replaced transaction (speed up or cancel)
+ * @param {Number} safeHeightBeforeEthTx Lower search bound
+ * @param {*} txInfo Transaction information to find the replacement
+ * @param {*} tx.nonce Nonce of the transaction searched
+ * @param {*} tx.from Signer of the transaction searched
+ * @param {*} tx.to Address called by the transaction searched
+ * @param {*} event.name Name of event expected if tx was speed up
+ * @param {*} event.abi tx.to contract abi needed to find event
+ * @param {*} event.validate Function to validate the content of event
+ */
+export async function findReplacementTx (safeHeightBeforeEthTx, tx, event) {
+  const web3 = new Web3(getEthProvider())
+
+  const currentNonce = await web3.eth.getTransactionCount(tx.from, 'latest')
+
+  // Transaction still pending
+  if (currentNonce <= tx.nonce) return null
+
+  // Binary search the block of replacement (canceled or speedup) transaction
+  // between safeHeightBeforeEthTx and latest.
+  let replacementTxBlock
+  let maxBlock = await web3.eth.getBlockNumber() // latest: chain head
+  let minBlock = safeHeightBeforeEthTx
+  while (minBlock <= maxBlock) {
+    const middleBlock = Math.floor((minBlock + maxBlock) / 2)
+    const middleNonce = await web3.eth.getTransactionCount(tx.from, middleBlock) - 1
+    if (middleNonce < tx.nonce) {
+      // middleBlock was mined before the tx with broadcasted nonce, so take next block as lower bound
+      minBlock = middleBlock + 1
+    } else if (middleNonce >= tx.nonce) {
+      // The middleBlock was mined after the tx with tx.nonce, so check if the account has a
+      // lower nonce at previous block which would mean that tx.nonce was mined in this middleBlock.
+      if (await web3.eth.getTransactionCount(tx.from, middleBlock - 1) - 1 < tx.nonce) {
+        // Confirm the nonce changed by checking the previous block:
+        // use previous block nonce `>=` broadcasted nonce in case there are multiple user tx
+        // in the previous block. If only 1 user tx, then `===` would work
+        replacementTxBlock = middleBlock
+        break
+      }
+      // Otherwise take the previous block as the higher bound
+      maxBlock = middleBlock - 1
+    }
+  }
+  if (!replacementTxBlock) {
+    const error = 'Could not find replacement transaction. It may be due to a chain reorg.'
+    throw new Error(error)
```

Comment:
It will throw this if I closed the tab, made my speed-up transaction, wait more than 20 blocks, and then re-open the tab? Should we search back to `tx.blockHeight` instead?

Reply:
It wouldn't throw in the scenario you describe. This throws in the very very unlikely case when the replacement tx got mined in a block that is 20 blocks before the original tx.

Reply:
Oh I see now and I'm not sure why I was confused. Starting `minBlock` is 20 blocks _before_ `tx.blockHeight`.